src/natcap/invest/managed_raster/managed_raster.pyx

```diff
@@ -300,3 +310,118 @@ cdef class _ManagedRaster:
         if self.write_mode:
             raster_band = None
             raster = None
+
+
+cdef class ManagedFlowDirRaster(_ManagedRaster):
+
+    cdef bint is_local_high_point(self, long xi, long yi):
+        """Check if a given pixel is a local high point.
+
+        Args:
+            xi (int): x coordinate of the pixel to consider
+            yi (int): y coordinate of the pixel to consider
+
+        Returns:
+            True if the pixel is a local high point, i.e. it has no
+            upslope neighbors; False otherwise.
+        """
+        ns = list(self.yield_upslope_neighbors(xi, yi))
+        if ns:
+            return False
+        return True
+
+    @cython.cdivision(True)
+    cdef vector[NeighborTuple] yield_upslope_neighbors(
+            ManagedFlowDirRaster self, long xi, long yi):
+        """Yield upslope neighbors of a given pixel.
+
+        Args:
+            xi (int): x coordinate of the pixel i to consider
+            yi (int): y coordinate of the pixel i to consider
+
+        Yields:
+            (j, xj, yj, p_ij) tuples representing neighboring pixels, where
+            j is the index 0-7 of the neighbor,
+            xj is the x coordinate of pixel j in pixel space,
+            yj is the y coordinate of pixel j in pixel space, and
+            p_ij is the proportion of flow from pixel j that flows into pixel i
+        """
+        cdef int n_dir, flow_dir_j, idx
+        cdef long xj, yj
+        cdef float flow_ji, flow_dir_j_sum
+
+        cdef NeighborTuple n
+        cdef vector[NeighborTuple] upslope_neighbor_tuples
+
+        for n_dir in xrange(8):
+            xj = xi + COL_OFFSETS[n_dir]
+            yj = yi + ROW_OFFSETS[n_dir]
+            if (xj < 0 or xj >= self.raster_x_size or
+                    yj < 0 or yj >= self.raster_y_size):
+                continue
+            flow_dir_j = <int>self.get(xj, yj)
+            flow_ji = (0xF & (flow_dir_j >> (4 * FLOW_DIR_REVERSE_DIRECTION[n_dir])))
+
+            if flow_ji:
+                flow_dir_j_sum = 0
+                for idx in range(8):
+                    flow_dir_j_sum += (flow_dir_j >> (idx * 4)) & 0xF
+                n.direction = n_dir
+                n.x = xj
+                n.y = yj
+                n.flow_proportion = flow_ji / flow_dir_j_sum
+                upslope_neighbor_tuples.push_back(n)
+
+        return upslope_neighbor_tuples
+
+    @cython.cdivision(True)
+    cdef vector[NeighborTuple] yield_downslope_neighbors(
+            ManagedFlowDirRaster self, long xi, long yi, bint skip_oob=True):
+        """Yield downslope neighbors of a given pixel.
+
+        Args:
+            xi (int): x coordinate of the pixel i to consider
+            yi (int): y coordinate of the pixel i to consider
+            skip_oob (bool): if True, do not yield neighbors that fall
+                outside the raster bounds.
+
+        Yields:
```

Comment:
Can we update this to `Returns`? It's a very nitpicky thing but I have been wondering whether it's worth changing all the `yield` terminology used to `get` or `return`, since we're no longer yielding anything. In the Python sense anyway.

Reply:
Updated to `Returns`